master/internal/stream/system.go

```diff
@@ -324,28 +436,48 @@ func doPublishLoop[T stream.Msg](
 			}
 			// Broadcast all the events.
 			publisher.Broadcast(events)
-			break
 		}
 	}
+}
 
-	return nil
+func newPermFilter[T stream.Msg](
+	ctx context.Context,
+	user model.User,
+	permFilterFn func(context.Context, model.User) (func(T) bool, error),
+	err *error,
+) func(T) bool {
+	out, tempErr := permFilterFn(ctx, user)
+	*err = tempErr
```

Comment:
whoa, you are unconditionally overwriting the old error here, that's gonna be a really sneaky bug.

Anytime you want to use a monadic error like this, you need to first check if the error is already set, and do nothing in that case.